Bump WooCommerce minimum required PHP version to 7.2

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Note for reviewers

It's recommended to review each commit individually as they are pretty much orthogonal. Also although there are 100+ changed files, most of the changes are in two of the unit test related commits and are quite repetitive code additions or replacements.

### Changes proposed in this Pull Request

[As announced back in January](https://developer.woocommerce.com/2022/01/05/new-requirement-for-woocommerce-6-5-php-7-2/) the minimum PHP version required will be 7.2 as of WooCommerce 6.5. This pull request makes effective this transition and adjusts dependencies and unit tests as needed. Note that there's also a breaking change related to the Emogrifier package and the `woocommerce_emogrifier` hook.

#### Core changes and package updates

* _Requires PHP_ is changed in `woocommerce.php`
* _require-php_ and _config-platform-php_ is changed in `composer.json`
* Dependency versions update in `composer.json`:
  * The _pelago/emogrifier_ package version is changed to `^6.0`
  * The _maxmind-db/reader_ package version is bumped to `^1.6`
  * The _composer/installers_ package version is bumped to `^1.9` (a bump to 2.0 not possible due to conflict with Admin package)
  * The required version for _phpunit_ is now explicitly indicated as 7.5.20, otherwise it will install v8 as it's the newerst version that supports PHP 7.2.

#### Changes in unit tests

Some of the changes are for making the unit tests ready for PHPUnit 8, which is what WooCommerce will move to once the minimum supported WordPress version is 5.9 ([details here](https://make.wordpress.org/core/handbook/references/phpunit-compatibility-and-wordpress-versions/)).

* PHP 7.0 and 7.1 are removed from the list of PHP versions to run the unit tests in CI.
* Add explicit `void` return types to `setUp` and `tearDown` methods in unit test classes.
* Fix calls to deprecated PHPUnit methods, e.g. `assertInternalType` are replaced with `assertIs(type)`.
* Add [even more](https://github.com/woocommerce/woocommerce/pull/31286) `#[\ReturnTypeWillChange]` attributes, these are required for PHP 8.1.

#### Breaking change

WooCommerce has a `woocommerce_emogrifier` action that can be used to tweak how the emails WooCommerce sends are styled. Until now this action got an object of type `Emogrifier` as a parameter, the problem is that as of v4 of the Emogrifier package this class no longer exists, and in order to be compatible with PHP 8 we need at least v5 of the package.

The workaround in this pull request is to pass an instance of the `CssInliner` class, which is the replacement for the `Emogrifier` class. The new class keeps some of the methods of the old class but not all of them, hence the breaking change.

For reference this is the list of the `Emogrifier` methods that are still present in `CssInliner`:

* addAllowedMediaType
* addExcludedSelector
* disableInlineStyleAttributesParsing
* removeAllowedMediaType
* removeExcludedSelector
* setDebug

#### Other changes

A visibility modifier (`public` or `private`) has been added to the handful of constants defined in the classes in `src`, visibility modifier for constants is a feature introduced in PHP 7.2 that improves the code readability. The code in `src/Admin` and `src/Internal/Admin`  has been excluded frrom this refactor since there's a lot of files here and that deserves a separate project.

### How to test the changes in this Pull Request:

Test that WooCommerce works and the unit tests still pass (TODO: this could probably be improved)

One of the issues that triggered the bump to PHP 7.2 was [the impossibility to install WooCommerce as a Composer package in PHP 8](https://github.com/woocommerce/woocommerce/issues/30027). To test that this has been fixed:

1. Create an empty directory with any name.
2. Inside that directory create a `composer.json` file with this content (adjust "url" to the location of your local copy of WooCommerce, but remember that it must end with `/plugins/woocommerce`):

```json
{
    "name": "foobar/fizzbuzz",
    "minimum-stability": "dev",
    "repositories": [
      {
          "type": "path",
          "url": "../woocommerce/plugins/woocommerce"
      }
    ],
    "require": {
        "woocommerce/woocommerce": "*"
    }
}
```

3. Verify that `composer install` succeeds for all versions of PHP from 7.2 to 8.1 when your local copy of WooCommerce is in the branch of this pull request (in `trunk` it will fail in PHP 8.x).

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Tweak - Bump minimum required PHP version of WooCommerce to 7.2

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Bump WooCommerce minimum required PHP version to 7.2

This includes:

- Changing "Requires PHP" in woocommerce.php
- Changing "require-php" and "config-platform-php" in composer.json
- Bumping pelago/emogrifier to ^6.0 in composer.json
- Bumping maxmind-db/reader to ^1.6 in composer.json
- Bumping composer/installers to ^1.9 in composer.json
  (bump to 2.0 not possible due to conflict with Admin package)
- Removing jobs for 7.0/.1 unit test runs in CI
- Explicitly specifying the PHPUnit version to be 7.5,
  otherwise it will install v8 which isn't compatible with
  WordPress <5.9

Konamiman committed:
Add return type setUp and tearDown methods in unit tests

After the bump to PHP 7.2 all the setUp and tearDown methods
in unit test classes need to have an explicit "void" return type
declaration.

Konamiman committed:
Add .phpunit.result.cache to .gitignore

This file is created by PHPUnit v8 when tests run.

Konamiman committed:
Fix calls to deprecated PHPUnit methods in unit tests.

Fix tests where PHPUnit 8 would throw "this method is deprecated
and will be removed in PHPUnit 9, use this other one instead"
warnings, by replacing all those with the proper ones. This will help
once we are able to use PHP 8 (once WP requirement is bumped to 5.9)

Konamiman committed:
Use CssInliner class instead of Emogrifier class

With the upgrade of the Emogrifier package we can't use the Emogrifier
class, since it was removed in v4 (we need to use at least v5 for
compatibility with PHP 8); so we now use the CssInliner class.

Note that this is a potential breaking change for consumers of the
woocommerce_emogrifier hook, which was getting an Emogrifier object
as parameter (now it gets a CssInliner object, which is partially
compatible).

Konamiman committed:
Add more "#[\ReturnTypeWillChange]" attributes.

These are needed for the code to be compatible with both PHP 7.x and 8.x

Konamiman committed:
Add visibility modifiers to constants in src

PHP 7.2 allows visibility modifiers in constants, so let's add these
to the code in the src directory.

The src/Admin and src/Internal/Admin directories are skipped as
they contain a lot of code and they deserve a separate project.

barryhughes commented on plugins/woocommerce/includes/emails/class-wc-email.php:
```diff
@@ -559,18 +563,20 @@ public function style_inline( $content ) {
+					do_action( 'woocommerce_emogrifier', $css_inliner, $this );
 
-					do_action( 'woocommerce_emogrifier', $emogrifier, $this );
+					$domDocument = $css_inliner->getDomDocument();
```
```suggestion
					$dom_document = $css_inliner->getDomDocument();
```

Konamiman replied:
The consequences of reckless copypaste. Fixed!

barryhughes commented on plugins/woocommerce/includes/emails/class-wc-email.php:
```diff
@@ -559,18 +563,20 @@ public function style_inline( $content ) {
-					$html_prune = \Pelago\Emogrifier\HtmlProcessor\HtmlPruner::fromHtml( $content );
-					$html_prune->removeElementsWithDisplayNone();
-					$content    = $html_prune->render();
+					HtmlPruner::fromDomDocument( $domDocument )->removeElementsWithDisplayNone();
```
```suggestion
					HtmlPruner::fromDomDocument( $dom_document )->removeElementsWithDisplayNone();
```

barryhughes commented on plugins/woocommerce/composer.json:
```diff
@@ -13,19 +13,20 @@
-		"maxmind-db/reader": "1.6.0",
-		"pelago/emogrifier": "3.1.0",
+		"composer/installers": "^1.9",
+		"maxmind-db/reader": "^1.6",
```
Regarding the installed version, at time of writing it will be 1.11.0 (which shares the [same min PHP requirements](https://github.com/maxmind/MaxMind-DB-Reader-php/blob/v1.11.0/composer.json#L16)). Worth bumping here, too?

```suggestion
		"maxmind-db/reader": "^1.11",
```

(Not essential, just a thought.)

Konamiman replied:
Makes sense, changed.

barryhughes reviewed (changes requested):
Flagged a couple of minor things, but looks good and works well.

Confirmed WC as a dependency works (at least, there are no errors when running `composer install`) with PHP versions 7.2 through to 8.1.

barryhughes commented:
Should we issue a dev advisory in relation to this change, and particularly regarding Emogrifier?

Konamiman committed:
Small changes related to the bump of min PHP to 7.2

- Rename wrongly cased variable
- Require higher version of maxmind/dbreader

Konamiman committed:
Merge branch 'trunk' into bump-required-php-to-7.2

Konamiman committed:
Add return type setUp and tearDown methods in admin unit tests

After the bump to PHP 7.2 all the setUp and tearDown methods
in unit test classes need to have an explicit "void" return type
declaration. This commit adds those for the woocommerce-admin tests.

Konamiman commented:
@barryhughes Apart from fixed the small things you found, I've added an extra commit to adjust the Admin unit tests that have been recently moved to the core unit tests suite.

barryhughes reviewed (approved):
LGTM.
